lib/sigstore/signer.rb

```diff
@@ -109,35 +122,59 @@ def fetch_cert(csr)
 
       resp_body = JSON.parse(resp.body)
 
-      chain = resp_body.fetch("signedCertificateEmbeddedSct").fetch("chain")
-                       .fetch("certificates").map { |pem| Internal::X509::Certificate.read(pem) }
+      cert = resp_body.fetch("signedCertificateEmbeddedSct").fetch("chain")
+                      .fetch("certificates").first.then { |pem| Internal::X509::Certificate.read(pem) }
       logger.debug { "Fetched cert from fulcio" }
-      chain
+      cert
     end
 
-    def verify_chain(chain)
+    def verify_chain(leaf)
       # Perform certification path validation (RFC 5280 §6) of the returned certificate chain with the pre-distributed
       # Fulcio root certificate(s) as a trust anchor.
 
       x509_store = OpenSSL::X509::Store.new
-      chain = chain.map(&:openssl)
-      leaf = chain.shift
-      root = chain.pop # TODO: use root from trusted root as trust anchor
-      x509_store.add_cert root
-      raise x509_store.error_string unless x509_store.verify(leaf, chain)
+      expected_chain = @trusted_root.fulcio_cert_chain
+
+      x509_store.add_cert expected_chain.last.openssl
+      unless x509_store.verify(leaf.openssl, expected_chain[..-2].map(&:openssl))
+        raise Error::Signing, "returned certificate does not validate: #{x509_store.error_string}"
+      end
+
+      chain = x509_store.chain
+      chain.shift # remove the leaf cert
+      chain.map! { |cert| Internal::X509::Certificate.new(cert) }
 
       logger.debug { "verified chain" }
 
-      # TODO: verify SCT in leaf
       # Extract a SignedCertificateTimestamp, which may be embedded as an X.509 extension in the leaf certificate or
       # attached separately in the SigningCertificate returned from the Identity Service.
       # Verify this SignedCertificateTimestamp as in RFC 9162 §8.1.3, using the root certificate from
       # the Certificate Transparency Log.
+      if (result = Verifier.verify_scts(leaf, chain, @verifier.rekor_client.ct_keyring)) &&
+         !result.verified?
+        raise Error::Signing, "Failed to verify SCTs: #{result.reason}"
+      end
 
-      # TODO: verify leaf subject
       # Check that the leaf certificate contains the subject from the certificate signing request and encodes the
       # appropriate AuthenticationServiceIdentifier in an extension with OID 1.3.6.1.4.1.57264.1.8.
 
+      fulcio_issuer = leaf.extension(Internal::X509::Extension::FulcioIssuer)
+      unless fulcio_issuer && fulcio_issuer.issuer == @identity_token.issuer
+        raise Error::Signing, "certificate does not contain expected Fulcio issuer"
+      end
+
+      unless leaf.subject.to_a.empty?
+        raise Error::Signing,
+              "certificate contains unexpected subject #{leaf.subject.to_a}"
+      end
+
+      general_names = leaf.extension(Internal::X509::Extension::SubjectAlternativeName).general_names
+      expected_san = [[:uniformResourceIdentifier, @identity_token.identity]]
```

Comment:
is this generally going to be true, even for other issuers?

Reply:
Nope, unfortunately -- the Google and GitHub (user, not actions) issuers will embed a SAN with an `rfc822Name` instead, corresponding to the underlying email subject.

IIRC the way we handle this on `sigstore-python` is to enforce that there's exactly one SAN and check the SAN's value, regardless of its specified type. This isn't ideal from a domain separation perspective but it's about as clean as it currently gets without a stronger notion of subject "type" on the OIDC side.

Reply:
hmm -- given the spec calls for the client to validate the returned SAN, it would be helpful if we could publish a simple mapping of ID providers to subject names somewhere in fulcio that clients can grab and substitute into?

Reply:
Yeah, this would be good to have -- I think the closest thing to this currently is https://github.com/sigstore/fulcio/blob/b2186c01da1ddf807bde3ea8c450226d8e001d88/pkg/config/config.go#L182-L201, but that's only in Fulcio's code and not documented anywhere separately AFAICT.

(IMO we could make this general in a follow-up, or weaken the check to not include the SAN type and only match on the SAN value. Your call!)